### PR TITLE
Smearing optimize passed

### PR DESCRIPTION
**Pull request by nohayassin**

### CODE REVIEW

File: src/algo/depth-to-rgb-calibration/debug.h [hunk 1]
```diff
@@ -10,6 +10,6 @@
 #define AC_LOG_PREFIX "AC1: "
 #define AC_LOG(TYPE,MSG) LOG_##TYPE( AC_LOG_PREFIX << (std::string)( to_string() << MSG ))
 //#define AC_LOG(TYPE,MSG) LOG_ERROR( AC_LOG_PREFIX << MSG )
-//#define AC_LOG(TYPE,MSG) std::cout << (std::string)( to_string() << "-" << #TYPE [0] << "- " << MSG ) << std::endl; //LOG_INFO((std::string)( to_string() << "-" << #TYPE [0] << "- " << MSG ));
+//#define AC_LOG(TYPE,MSG) std::cout << (std::string)( to_string() << std::setprecision(6)<<std::fixed << "-" << #TYPE [0] << "- " << MSG ) << std::endl; //LOG_INFO((std::string)( to_string() << "-" << #TYPE [0] << "- " << MSG ));
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/debug.h]: Please remove

File: src/algo/depth-to-rgb-calibration/frame-data.h [hunk 2]
```diff
@@ -28,18 +28,18 @@ namespace depth_to_rgb_calibration {
...
         std::vector<double> gradient_x;
         std::vector<double> gradient_y;
         std::vector< byte > section_map_depth;
-        std::vector<uint8_t> valid_edge_pixels_by_ir;
+        std::vector<double> valid_edge_pixels_by_ir;
         std::vector<byte> valid_section_map;
         std::vector<double>valid_gradient_x;
         std::vector<double>valid_gradient_y;
-        std::vector<double> edges2;
+        //std::vector<double> edges;
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/frame-data.h]: Remove if not needed

File: src/algo/depth-to-rgb-calibration/frame-data.h [hunk 3]
```diff
@@ -51,8 +51,9 @@ namespace depth_to_rgb_calibration {
         std::vector<double> local_region[4];
         std::vector<double> local_region_x[4];//for debug
         std::vector<double> local_region_y[4]; // for debug
-        std::vector<double> local_edges;
+        //std::vector<double> ir_edges;
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/frame-data.h]: Remove if not needed

File: src/algo/depth-to-rgb-calibration/frame-data.h [hunk 4]
```diff
@@ -67,16 +68,43 @@ namespace depth_to_rgb_calibration {
         std::vector<double> edges;
         std::vector<double> supressed_edges;                
         size_t n_strong_edges;
-        std::vector<direction> directions;
+        //std::vector<direction> valid_directions;
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/frame-data.h]: Remove if not needed

File: src/algo/depth-to-rgb-calibration/frame-data.h [hunk 5]
```diff
@@ -67,16 +68,43 @@ namespace depth_to_rgb_calibration {
...
         std::vector<double> subpixels_x;
         std::vector<double> subpixels_y;
-        std::vector< uint16_t> closest;
+        //std::vector< uint16_t> closest;
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/frame-data.h]: Remove if not needed

File: src/algo/depth-to-rgb-calibration/frame-data.h [hunk 6]
```diff
@@ -67,16 +68,43 @@ namespace depth_to_rgb_calibration {
...
         std::vector<double> weights;
         std::vector<double> direction_deg;
-        std::vector<double3> vertices;
+        std::vector<double3> vertices_all; //vertices_all
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/frame-data.h]: I thought you said it's not needed? In any case, remove the comment

File: src/algo/depth-to-rgb-calibration/frame-data.h [hunk 7]
```diff
@@ -67,16 +68,43 @@ namespace depth_to_rgb_calibration {
...
+        std::vector< byte > section_map_depth_inside;
+        std::vector<double> local_x;
+        std::vector<double> local_y;
+        std::vector<double> gradient;
+        std::vector<double> local_values;
+        std::vector<double> grad_in_direction;
+        std::vector<double> grad_in_direction_inside;
+        std::vector<double> values_for_subedges;
+        std::vector<double> closest;
+        //std::vector< byte > supressed_edges;
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/frame-data.h]: Remove if not needed

File: src/algo/depth-to-rgb-calibration/frame-data.h [hunk 8]
```diff
@@ -67,16 +68,43 @@ namespace depth_to_rgb_calibration {
...
+        std::vector<byte> valid_section_map;
+        std::vector<double> local_rc_subpixel;
+        std::vector<double> edge_sub_pixel;
+        std::vector<double> valid_directions;
+        std::vector<double> directions;
+        std::vector<double> valid_edge_sub_pixel;
+        std::vector<double > valid_edge_sub_pixel_x;
+        std::vector<double > valid_edge_sub_pixel_y;
+        std::vector<double> sub_points;
+        //std::vector<double3> vertices;
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/frame-data.h]: Remove

File: src/algo/depth-to-rgb-calibration/optimizer.cpp [hunk 9]
```diff
@@ -435,7 +435,118 @@ void grid_xy(
         }
     }
 }
+template<class T>
+std::vector<double> interpolation(std::vector<T> const &grid_points, std::vector<double> x[], std::vector<double> y[], int dim, int valid_size, int valid_width)
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.cpp:277]: If possible, I'd like you to do what we did with valid-scene.cpp: put all the smearing-specific code in preprocess.cpp

File: src/algo/depth-to-rgb-calibration/optimizer.cpp [hunk 10]
```diff
@@ -435,7 +435,118 @@ void grid_xy(
...
+        {
+            res.push_back(0);
+        }
+    }
+    return res;
+}
+std::vector<double> find_local_values_min(std::vector<double>& local_values)
+{
+    std::vector<double> res;
+    size_t size = local_values.size() / 4;
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.cpp:374]: You can resize res if you know the size in advance, then just assign to res[i]

File: src/algo/depth-to-rgb-calibration/optimizer.cpp [hunk 11]
```diff
@@ -435,7 +435,118 @@ void grid_xy(
...
+        auto val2 = *(it + 1);
+        auto val3 = *(it + 2);
+        auto val4 = *(it + 3);
+        it += 4;
+        double res_val = std::min(val1, std::min(val2, std::min(val3, val4)));
+        res.push_back(res_val);
+    }
+    return res;
+}
 void optimizer::set_depth_data(
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.cpp]: You'll have to explain it to me: why are we setting the ir_data in set_depth_data? Did you remove set_ir_data()?

Comment (maloel) [reply src/algo/depth-to-rgb-calibration/optimizer.cpp]: Also, if we're keeping set_ir_data(), then I'd like to keep the APIs with the same names as before, i.e. set_z_data()

File: src/algo/depth-to-rgb-calibration/optimizer.cpp [hunk 12]
```diff
@@ -536,81 +647,242 @@ void optimizer::set_depth_data(
...
+    /*xim = xim(isInside);
+    yim = yim(isInside); 
+    zValuesForSubEdges = zValuesForSubEdges(isInside);
+    zGradInDirection = zGradInDirection(isInside);
+    directionIndex = directionIndex(isInside);
+    weights = weights(isInside);
+    vertices = vertices(isInside,:);
+    sectionMapDepth = sectionMapDepth(isInside);*/
+    std::vector<double> weights;
+    for (auto i = 0; i < _z.is_inside.size(); i++) {
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.cpp:703]: Let's keep consistent code formatting: please put the brace on its own line

File: src/algo/depth-to-rgb-calibration/optimizer.cpp [hunk 13]
```diff
@@ -897,28 +1169,28 @@ double get_min(double x, double y)
...
-    {
-        if (z_data.supressed_edges[i])
-            z_data.weights.push_back(
-                get_min(get_max(z_data.supressed_edges[i] - _params.grad_z_min, (double)0),
-                    _params.grad_z_max - _params.grad_z_min));
-    }
-
-    return res;
-}
+//std::vector<double> optimizer::calculate_weights(z_frame_data& z_data)
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.cpp:1004]: Remove if no longer needed

File: src/algo/depth-to-rgb-calibration/optimizer.h [hunk 14]
```diff
@@ -63,6 +63,17 @@ namespace depth_to_rgb_calibration {
         double move_thresh_pix_val = 20;
         double move_threshold_pix_num = 62.2080;
 
+        //smearing
+        double max_sub_mm_z = 4;
+        double constant_weights = 1000;
+        double k_depth[3][3] = { {731.27344,0,529.27344 },{0,731.97656,402.32031},{0,0,1} };
+        //double k_depth_pinv_trans[3][3] = { { 0.0013674778, -1.1641532e-10,	2.3283064e-10 }, //pinv(params.Kdepth)';
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.h]: Try to remove extra comments

File: src/algo/depth-to-rgb-calibration/optimizer.h [hunk 15]
```diff
@@ -96,10 +107,10 @@ namespace depth_to_rgb_calibration {
...
             size_t width, size_t height );
-        void set_z_data(
+        /*void set_z_data(
             std::vector< z_t > && z_data,
             rs2_intrinsics_double const & depth_intrinsics,
-            float depth_units );
+            float depth_units );*/
         void set_depth_data(
             std::vector< z_t >&& z_data,
             std::vector< ir_t >&& ir_data,
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.h]: See comment elsewhere

File: src/algo/depth-to-rgb-calibration/optimizer.h [hunk 16]
```diff
@@ -141,7 +152,7 @@ namespace depth_to_rgb_calibration {
         std::vector<uint8_t> get_luminance_from_yuy2( std::vector<uint16_t> yuy2_imagh );
 
         std::vector<uint8_t> get_logic_edges( std::vector<double> edges );
-        std::vector<double> calculate_weights( z_frame_data& z_data );
+        //std::vector<double> calculate_weights( z_frame_data& z_data );
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.h]: Remove if not needed

File: src/algo/depth-to-rgb-calibration/optimizer.h [hunk 17]
```diff
@@ -167,8 +178,8 @@ namespace depth_to_rgb_calibration {
         params _params;
         yuy2_frame_data _yuy;
         ir_frame_data _ir;
+        //z_frame_data _z;
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.h]: Remove

File: src/depth-to-rgb-calibration.cpp [hunk 18]
```diff
@@ -70,8 +70,9 @@ depth_to_rgb_calibration::depth_to_rgb_calibration(
     AC_LOG( DEBUG, "... setting z data" );
     auto z_profile = depth.get_profile().as< rs2::video_stream_profile >();
     auto z_data = (impl::z_t const *) depth.get_data();
-    _algo.set_z_data(
+    _algo.set_depth_data(
```
Comment (maloel) [inline src/depth-to-rgb-calibration.cpp]: Shouldn't need changing

File: unit-tests/algo/depth-to-rgb-calibration/compare-scene.h [hunk 19]
```diff
@@ -48,45 +49,75 @@ void compare_scene( std::string const & scene_dir )
...
+    CHECK(compare_to_bin_file< double >(ir_data.gradient_y, scene_dir, "Iy",  z_w, z_h, "double_00", compare_same_vectors));
+    CHECK(compare_to_bin_file< double >(ir_data.edges, scene_dir, "iEdge",  z_w, z_h, "double_00", compare_same_vectors));
+    CHECK(compare_to_bin_file< double >(depth_data.edges, scene_dir, "zedge",  z_w, z_h, "double_00", compare_same_vectors));
+    CHECK(compare_to_bin_file< uint8_t >(depth_data.section_map_depth, scene_dir, "sectionMapDepth",   z_w, z_h, "uint8_00",compare_same_vectors));
+    CHECK(compare_to_bin_file< double >(ir_data.valid_edge_pixels_by_ir, scene_dir, "validEdgePixelsByIR",  z_w, z_h, "double_00", compare_same_vectors));
+    CHECK(compare_to_bin_file< double >(ir_data.valid_location_rc_x, scene_dir, "gridXValid", 1, 105794,  "double_00", compare_same_vectors));
+    CHECK(compare_to_bin_file< double >(ir_data.valid_location_rc_y, scene_dir, "gridYValid", 1, 105794,  "double_00", compare_same_vectors));
+    CHECK(compare_to_bin_file< double >(ir_data.valid_location_rc, scene_dir, "locRC", 2, 105794,  "double_00", compare_same_vectors));
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/compare-scene.h]: No magic numbers please -- matlab should communicate the dimensions to C++ somehow

File: unit-tests/algo/depth-to-rgb-calibration/compare-scene.h [hunk 20]
```diff
@@ -29,13 +29,14 @@ void compare_scene( std::string const & scene_dir )
     auto z_w = ci.z.width;
     auto num_of_calib_elements = 32;
 
+#if 0
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/compare-scene.h]: Remove if no longer relevant

File: unit-tests/algo/depth-to-rgb-calibration/compare-scene.h [hunk 21]
```diff
@@ -95,8 +126,8 @@ void compare_scene( std::string const & scene_dir )
     // edge distribution
     CHECK( compare_to_bin_file< double >( z_data.sum_weights_per_section, scene_dir, "depthEdgeWeightDistributionPerSectionDepth", 1, 4, "double_00", compare_same_vectors ) );
 
-    CHECK( compare_to_bin_file< byte >( z_data.section_map, scene_dir, "sectionMapDepth_trans", 1, md.n_edges, "uint8_00", compare_same_vectors ) );
-    CHECK( compare_to_bin_file< byte >( yuy_data.section_map, scene_dir, "sectionMapRgb_trans", 1, rgb_w*rgb_h, "uint8_00", compare_same_vectors ) );
+    //CHECK( compare_to_bin_file< byte >( z_data.section_map, scene_dir, "sectionMapDepth_trans", 1, md.n_edges, "uint8_00", compare_same_vectors ) );
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/compare-scene.h:102]: Remove if not needed

File: unit-tests/algo/depth-to-rgb-calibration/compare-scene.h [hunk 22]
```diff
@@ -116,8 +147,8 @@ void compare_scene( std::string const & scene_dir )
     // 3. movement
     CHECK( compare_to_bin_file< double >( yuy_data.gaussian_diff_masked, scene_dir, "IDiffMasked", rgb_w, rgb_h, "double_00", compare_same_vectors ) );
     CHECK( compare_to_bin_file< uint8_t >( yuy_data.move_suspect, scene_dir, "ixMoveSuspect", rgb_w, rgb_h, "uint8_00", compare_same_vectors ) );
-
-
+#endif
+#if 1
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/compare-scene.h:124]: Remove

File: unit-tests/algo/depth-to-rgb-calibration/compare-scene.h [hunk 23]
```diff
@@ -170,8 +201,8 @@ void compare_scene( std::string const & scene_dir )
     auto cost = cal.get_cost();
 
     CHECK( compare_calib_to_bin_file( new_calibration, cost, scene_dir, "new_calib", num_of_calib_elements, 1, "double_00" ) );
-
-
+#endif
+#if 0
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/compare-scene.h:178]: Why?

File: unit-tests/algo/depth-to-rgb-calibration/test-scene-2.cpp [hunk 24]
```diff
@@ -10,9 +10,9 @@
 
 TEST_CASE("Scene 2", "[d2rgb]")
 {
-    std::string scene_dir( "..\\unit-tests\\algo\\depth-to-rgb-calibration\\19.2.20" );
+    std::string scene_dir("C:\\Users\\nyassin\\Documents\\realsense_all\\debug_scene");// "..\\unit-tests\\algo\\depth-to-rgb-calibration\\19.2.20" );
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/test-scene-2.cpp]: Undo

File: unit-tests/approx.h [hunk 25]
```diff
@@ -17,7 +17,7 @@
 //     REQUIRE( <val> == Approx(2.1).epsilon(0.01) );  // allow up to 1% diff
 // So we want a macro that does this automatically:
 #if ! defined( __EPSILON )
-#define __EPSILON (std::numeric_limits<float>::epsilon()*1)
+#define __EPSILON (std::numeric_limits<float>::epsilon())
```
Comment (maloel) [inline unit-tests/approx.h]: Undo

File: src/algo/depth-to-rgb-calibration/optimizer.h [hunk 26]
```diff
@@ -63,6 +63,13 @@ namespace depth_to_rgb_calibration {
         double move_thresh_pix_val = 20;
         double move_threshold_pix_num = 62.2080;
 
+        //smearing
+        double max_sub_mm_z = 4;
+        double constant_weights = 1000;
+        double k_depth[3][3] = { {731.27344,0,529.27344 },{0,731.97656,402.32031},{0,0,1} };
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.h]: These seem like low precision... they're a copy-paste from matlab, or manually entered?

What's this k_depth of, anyway?

Comment (maloel) [reply src/algo/depth-to-rgb-calibration/optimizer.h]: Please explain that this is the output of a constant expression run thru a function in algo.
Check if k_depth can be removed.

File: src/algo/depth-to-rgb-calibration/valid-scene.cpp [hunk 27]
```diff
@@ -702,13 +702,14 @@ bool optimizer::is_scene_valid()
 
     // remove pixels in section map that were removed in weights
     AC_LOG(DEBUG, "... " << _z.supressed_edges.size() << " total edges");
-    for (auto i = 0; i < _z.supressed_edges.size(); i++)
+    /*for (auto i = 0; i < _z.supressed_edges.size(); i++)
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/valid-scene.cpp]: Remove?

File: src/depth-to-rgb-calibration.cpp [hunk 28]
```diff
@@ -72,6 +72,7 @@ depth_to_rgb_calibration::depth_to_rgb_calibration(
     auto z_data = (impl::z_t const *) depth.get_data();
     _algo.set_z_data(
         std::vector< impl::z_t >( z_data, z_data + depth.get_data_size() / sizeof( impl::z_t ) ),
+        std::vector< impl::ir_t >(ir_data, ir_data + ir.get_data_size() / sizeof(impl::ir_t)),
```
Comment (maloel) [inline src/depth-to-rgb-calibration.cpp]: Not needed

File: src/algo/depth-to-rgb-calibration/valid-scene.cpp (hunk 27 above)
Comment (nohayassin) [reply src/algo/depth-to-rgb-calibration/valid-scene.cpp]: it is related to scene validity, I will keep it for now

File: src/algo/depth-to-rgb-calibration/optimizer.h [hunk 29]
```diff
@@ -130,8 +131,8 @@ namespace depth_to_rgb_calibration {
         z_frame_data    const & get_z_data() const   { return _z; }
         yuy2_frame_data const & get_yuy_data() const { return _yuy; }
         ir_frame_data   const & get_ir_data() const  { return _ir; }
-        z_frame_data    const& get_depth_data() const { return _depth; }
-        // impl
+        z_frame_data    const& get_depth_data() const { return _z; }
```
Comment (maloel) [inline src/algo/depth-to-rgb-calibration/optimizer.h]: Not needed

File: unit-tests/algo/depth-to-rgb-calibration/CMakeLists.txt [hunk 30]
```diff
@@ -102,3 +91,67 @@ dl_algo_file( "${F9440687}/2/binFiles/xCoeff_P_5_5089x12_double_00.bin" "")
 dl_algo_file( "${F9440687}/2/binFiles/xCoeff_P_6_5089x12_double_00.bin" "")
 
 dl_algo_file( "${F9440687}/2/binFiles/camera_params" "")
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/CMakeLists.txt]: Shouldn't this file move to ac1x/?

File: unit-tests/algo/depth-to-rgb-calibration/scene-data.h [hunk 31]
```diff
@@ -108,6 +108,8 @@ struct scene_metadata
     uint64_t n_iterations;        // how many steps through optimization, and how many iteration file sets
     double correction_in_pixels;  // XY movement
     uint64_t n_edges;             // strong edges, i.e. after suppression
+    uint64_t valid_pixels;
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/scene-data.h]: It's the "number of" pixel and not the pixels themselves... please denote this with "n_" here and for valid_ir_edges.

File: unit-tests/algo/depth-to-rgb-calibration/test-scene-2.cpp (hunk 24 above)
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/test-scene-2.cpp]: Undo :)

File: unit-tests/algo/depth-to-rgb-calibration/test-scene-2.cpp [hunk 32]
```diff
@@ -10,9 +10,9 @@
...
 //    std::string scene_dir( "C:\\work\\autocal" );
-    scene_dir += "\\F9440687\\LongRange_D_768x1024_RGB_1920x1080\\2\\";
+    scene_dir += "\\LongRange 768X1024 (RGB 1920X1080)\\2\\";// "\\F9440687\\LongRange_D_768x1024_RGB_1920x1080\\2\\";
```
Comment (maloel) [inline unit-tests/algo/depth-to-rgb-calibration/test-scene-2.cpp]: Undo

File: unit-tests/algo/depth-to-rgb-calibration/CMakeLists.txt (hunk 30 above)
Comment (nohayassin) [reply unit-tests/algo/depth-to-rgb-calibration/CMakeLists.txt]: we need 2 of them: in ac1x and under binFiles

**Pull request merged**